ddtrace/llmobs/_integrations/google_genai_utils.py

```diff
@@ -0,0 +1,161 @@
+    """
+    contents has a complex union type structure:
+    - contents: Union[ContentListUnion, ContentListUnionDict]
+    - ContentListUnion = Union[list[ContentUnion], ContentUnion]
+    - ContentListUnionDict = Union[list[ContentUnionDict], ContentUnionDict]
+    - ContentUnion = Union[Content, list[PartUnion], PartUnion]
+    - PartUnion = Union[File, Part, str]
+
+    Can also be used for system_instruction which has type ContentUnion
+
+    This function normalizes all these variants into a list of dicts with format{"role": role, "parts": parts}
+    """
+
+    def extract_content(content):
+        role = _get_attr(content, "role", None)
+        parts = _get_attr(content, "parts", None)
+
+        # if parts is missing and content itself is a PartUnion or list[PartUnion]
+        if parts is None:
+            if isinstance(content, list):
+                parts = content
+            else:
+                parts = [content]
+
+        elif not isinstance(parts, list):
+            parts = [parts]
+
+        return {"role": role, "parts": parts}
+
+    if isinstance(contents, list):
+        return [extract_content(c) for c in contents]
+    return [extract_content(contents)]
+
+
+def extract_metrics_google_genai(response) -> Dict[str, Any]:
+    if not response:
+        return {}
+
+    usage_metadata = _get_attr(response, "usage_metadata", {})
+
+    usage = {}
+    input_tokens = _get_attr(usage_metadata, "prompt_token_count", None)
+    output_tokens = _get_attr(usage_metadata, "candidates_token_count", None)
+    cached_tokens = _get_attr(usage_metadata, "cached_content_token_count", None)
+    total_tokens = _get_attr(usage_metadata, "total_token_count", None) or input_tokens + output_tokens
+
+    if input_tokens is not None:
+        usage[INPUT_TOKENS_METRIC_KEY] = input_tokens
+    if output_tokens is not None:
+        usage[OUTPUT_TOKENS_METRIC_KEY] = output_tokens
+    if cached_tokens is not None:
+        usage["cached_tokens"] = cached_tokens  # TODO(max): add constant for cached tokens when it is introduced
+    if total_tokens is not None:
+        usage[TOTAL_TOKENS_METRIC_KEY] = total_tokens
+
+    return usage
+
+
+# TODO(max): check for cases where more than one field is set per part
+def extract_message_from_part_google_genai(part, role: str) -> Dict[str, Any]:
+    """part is a PartUnion = Union[File, Part, PIL_Image, str]
+
+    returns a dict representing a message with format {"role": role, "content": content}
+    """
+    # substitute "model" with "assistant" to stay consistent with other integrations
+    if role == "model":
+        role = DEFAULT_MODEL_ROLE
+
+    message: Dict[str, Any] = {"role": role}
+    if isinstance(part, str):
+        message["content"] = part
+        return message
+
+    thought = _get_attr(part, "thought", False)
+    if thought:
+        message["role"] = "reasoning"
+
+    text = _get_attr(part, "text", None)
+    if text:
+        message["content"] = text
```

Comment:
```suggestion
        message["content"] = str(text)
```
Let's cast these values to str if we're going to be setting them as `message["content"]`